File: code/zato-web-admin/src/zato/admin/static/js/security/wss.js

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.WSSecurity = new Class({
    toString: function() {
        var s = '<WSSecurity id:{0} name:{1} is_active:{2}>';
        return String.format(s, this.id ? this.id : '(none)',
                                this.name ? this.name : '(none)',
                                this.is_active ? this.is_active : '(none)');
    }
});

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.password_required = false;
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.WSSecurity;
    $.fn.zato.data_table.new_row_func = $.fn.zato.security.wss.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms([
        'name',
        'username',
    ]);
    $.fn.zato.data_table.before_submit_hook = $.fn.zato.security.wss.beforeSubmitHook;

    // .. the mode selector decides which tabs apply ..
    $(document).on('change', '#id_mode', function() {
        $.fn.zato.security.wss.update_mode_tabs('create');
    });
    $(document).on('change', '#id_edit-mode', function() {
        $.fn.zato.security.wss.update_mode_tabs('edit');
    });
})

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.config = {

    // Which tab holds the signing and encryption toggles, and what the form says
    // when an X.509 definition turns both of them off.
    'cryptoTab': 'crypto',
    'x509Mode': 'x509',
    'noX509OperationMessage': 'An X.509 definition needs signing, encryption or both'
};

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.tab_labels = {
    main:   'Main',
    token:  'Username token',
    saml:   'SAML',
    crypto: 'Crypto material'
};

$.fn.zato.security.wss._reset_tabs = function(action) {
    var is_edit = action === 'edit';
    $.fn.zato.form_tabs.reset({
        div_id:       is_edit ? '#edit-div' : '#create-div',
        panel_prefix: is_edit ? 'wss-edit-tab-panel-' : 'wss-create-tab-panel-',
        default_tab:  'main',
        tab_labels:   $.fn.zato.security.wss.tab_labels
    });
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Mode-dependent tab visibility
// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.update_mode_tabs = function(action) {

    var is_edit = action === 'edit';
    var div = $(is_edit ? '#edit-div' : '#create-div');
    var mode = $(is_edit ? '#id_edit-mode' : '#id_mode').val();

    // Which tabs the mode makes use of - Main is always there,
    // the crypto material applies to both X.509 and SAML.
    var visible_by_mode = {
        'username_token': {token: true,  saml: false, crypto: false},
        'x509':           {token: false, saml: false, crypto: true},
        'saml':           {token: false, saml: true,  crypto: true}
    };

    var visible = visible_by_mode[mode] || {};

    div.find('.dashboard-tab').each(function() {
        var tab = $(this).attr('data-tab');
        if(tab === 'main') {
            return;
        }
        var show = !!visible[tab];
        $(this).toggle(show);

        // A tab that just went away cannot stay active - fall back to Main.
        if(!show && $(this).hasClass('dashboard-tab-active')) {
            div.find('.dashboard-tab[data-tab="main"]').click();
        }
    });
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.beforeSubmitHook = function(form) {

    var config = $.fn.zato.security.wss.config;
    var isEdit = $(form).attr('id').indexOf('edit') !== -1;
    var prefix = isEdit ? 'edit-' : '';

    var signSelector = '#id_' + prefix + 'sign';
    var encryptSelector = '#id_' + prefix + 'encrypt';

    // Clear anything left over from a previous attempt ..
    $.fn.zato.remove_css_attention(signSelector);
    $.fn.zato.remove_css_attention(encryptSelector);

    // .. signing and encryption belong to the X.509 mode alone ..
    var mode = $('#id_' + prefix + 'mode').val();

    if(mode !== config.x509Mode) {
        return true;
    }

    // .. and that mode needs at least one of the two.
    if($(signSelector).is(':checked') || $(encryptSelector).is(':checked')) {
        return true;
    }

    // The two toggles sit on their own tab, which has to be the active one
    // before the form can point at them.
    var div = $(isEdit ? '#edit-div' : '#create-div');
    div.find('.dashboard-tab[data-tab="' + config.cryptoTab + '"]').click();

    $.fn.zato.draw_attention_to(signSelector);
    $.fn.zato.draw_attention_to(encryptSelector);
    $.fn.zato.show_bottom_tooltip(signSelector, config.noX509OperationMessage, true);

    return false;
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.field_descriptions = {

    // Main tab
    'id_name': 'A unique name for this definition. Used to identify it in logs and the dashboard.',
    'id_is_active': 'Whether this definition can be used. Connections do not apply inactive definitions.',
    'id_mode': 'The credential scheme this definition carries - a username token, ' +
        'X.509 signing and encryption or a SAML assertion.',
    'id_username': 'The username the definition operates with. With username tokens it goes into the message, ' +
        'with other modes it identifies the definition.',

    // Username token tab
    'id_use_digest': 'When on, the password goes out in digest form with a nonce and a timestamp ' +
        'instead of clear text. The profile mandates SHA-1 for the digest, and verifying one requires ' +
        'the password to be recoverable here, not stored as a hash.',

    // SAML tab
    'id_issuer': 'The entity that vouches for the assertion, e.g. the URI of your organization or gateway.',
    'id_subject': 'Who the assertion is about, e.g. the identity of the calling system or user.',
    'id_audience': 'Optionally, who the assertion is meant for. ' +
        'Leave empty if the endpoint does not restrict it.',

    // Crypto material tab
    'id_sign': 'When on, outgoing messages are signed - the message body with X.509, ' +
        'the assertion itself with SAML.',
    'id_encrypt': 'When on, the message body is encrypted for the recipient using their certificate. ' +
        'Applies to the X.509 mode.',
    'id_signing_key': 'Path to a PEM file on the server with your private key, used to produce signatures.',
    'id_signing_certificate_chain': 'Path to a PEM file on the server with the certificate matching ' +
        'the signing key, optionally followed by its chain. It travels with the message so the other side ' +
        'can verify the signature.',
    'id_decryption_key': 'Path to a PEM file on the server with the private key that decrypts ' +
        'incoming messages. With RSA it is usually the signing key again.',
    'id_peer_certificate': 'Path to a PEM file on the server with the other side\'s certificate - ' +
        'used to encrypt to them and to pin their signatures.',
    'id_trust_anchors': 'Path to a PEM file on the server with CA certificates that the other side\'s ' +
        'signing certificates may chain up to, as an alternative to pinning one certificate.',
};

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.create = function() {
    $.fn.zato.security.wss._reset_tabs('create');
    $.fn.zato.data_table._create_edit('create', 'Create a new WS-Security definition', null);
    $.fn.zato.security.wss.update_mode_tabs('create');
    $.fn.zato.how_it_works.init({
        badgeId: 'create-how-it-works',
        divId: '#create-div',
        descriptions: $.fn.zato.security.wss.field_descriptions
    });
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.edit = function(id) {
    $.fn.zato.security.wss._reset_tabs('edit');
    $.fn.zato.data_table._create_edit('edit', 'Update the WS-Security definition', id);
    $.fn.zato.security.wss.update_mode_tabs('edit');
    $.fn.zato.how_it_works.init({
        badgeId: 'edit-how-it-works',
        divId: '#edit-div',
        descriptions: $.fn.zato.security.wss.field_descriptions
    });
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss._to_django_bool = function(value) {
    return (value === true || value == 'on' || value == 'True') ? 'True' : 'False';
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.data_table.new_row = function(item, data, include_tr) {
    let row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    let is_active = $.fn.zato.security.wss._to_django_bool(item.is_active) == 'True';

    let mode_labels = {
        'username_token': 'Username token',
        'x509': 'X.509',
        'saml': 'SAML'
    };

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";

    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', is_active ? 'Yes' : 'No');

    row += String.format('<td>{0}</td>', item.username);
    row += String.format('<td>{0}</td>', mode_labels[item.mode] || item.mode);

    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.data_table.change_password('{0}')\">Change password</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.security.wss.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.security.wss.delete_('{0}');\">Delete</a>", item.id));
    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);

    row += String.format("<td class='ignore'>{0}</td>", item.is_active);
    row += String.format("<td class='ignore'>{0}</td>", item.mode);
    row += String.format("<td class='ignore'>{0}</td>", $.fn.zato.security.wss._to_django_bool(item.use_digest));
    row += String.format("<td class='ignore'>{0}</td>", $.fn.zato.security.wss._to_django_bool(item.sign));
    row += String.format("<td class='ignore'>{0}</td>", $.fn.zato.security.wss._to_django_bool(item.encrypt));
    row += String.format("<td class='ignore'>{0}</td>", item.issuer ? item.issuer : '');
    row += String.format("<td class='ignore'>{0}</td>", item.subject ? item.subject : '');
    row += String.format("<td class='ignore'>{0}</td>", item.audience ? item.audience : '');
    row += String.format("<td class='ignore'>{0}</td>", item.signing_key ? item.signing_key : '');
    row += String.format("<td class='ignore'>{0}</td>", item.signing_certificate_chain ? item.signing_certificate_chain : '');
    row += String.format("<td class='ignore'>{0}</td>", item.decryption_key ? item.decryption_key : '');
    row += String.format("<td class='ignore'>{0}</td>", item.peer_certificate ? item.peer_certificate : '');
    row += String.format("<td class='ignore'>{0}</td>", item.trust_anchors ? item.trust_anchors : '');

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.security.wss.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'WS-Security definition `{0}` deleted',
        'Are you sure you want to delete WS-Security definition `{0}`?',
        true);
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

File: code/zato-common/src/zato/common/soap/security/wss.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2026, Zato Source s.r.o. https://zato.io

Licensed under AGPLv3, see LICENSE.txt for terms and conditions.
"""

# stdlib
from pathlib import Path

# Zato
from zato.common.soap.common import NS, SOAPSecurityException
from zato.common.soap.security.saml import add_assertion, add_attribute, get_assertion, new_assertion, sign_assertion, \
    validate_assertion_conditions, verify_assertion
from zato.common.soap.security.usernametoken import add_username_token, verify_username_token
from zato.common.soap.security.x509 import decrypt_body, encrypt_body, sign, verify, VerifiedSignature
from zato.common.util.xml_.core import qname
from zato.common.util.xml_.keystore import load_certificates_pem, load_private_key_pem, new_keystore

# ################################################################################################################################
# ################################################################################################################################

if 0:
    from zato.common.typing_ import any_, anydict, stranydict
    from zato.common.util.xml_.keystore import Keystore
    any_ = any_
    anydict = anydict
    Keystore = Keystore
    stranydict = stranydict

# ################################################################################################################################
# ################################################################################################################################

class Mode:
    """ The modes a WS-Security definition can be in - which credential scheme
    the definition carries.
    """
    UsernameToken = 'username_token'
    X509 = 'x509'
    SAML = 'saml'

# All the modes a definition may use, in the order the Dashboard shows them.
All_Modes = (Mode.UsernameToken, Mode.X509, Mode.SAML)

# ################################################################################################################################
# ################################################################################################################################

# The keystores built so far, keyed by what identifies a definition's trust material. Entries are
# dropped when a definition is edited or deleted, so this grows with the number of definitions
# rather than with traffic.
_keystore_cache:'anydict' = {}

# What a definition that does not say so is taken to mean. All three are optional inputs stored in a
# definition's opaque attributes, so a definition that never set one simply has no such key - the
# dashboard always sends all three as real booleans, but an enmasse YAML definition carries only the
# keys it was written with, and reading them directly turned a plaintext-password definition without
# a use_digest line into a KeyError on every request, which surfaced to the caller as a 500 rather
# than as an authentication result.
#
# False is what the absence means in each case rather than a lenient guess. The UsernameToken profile
# reads an absent password type as clear text, and a definition that does not ask for a signature or
# for encryption has not configured anything for these to skip - a definition that does ask and has
# nothing to verify against still fails closed, in validate_certificate_chain.
Default_Use_Digest = False
Default_Sign = False
Default_Encrypt = False

# What an X.509 definition that asks for neither a signature nor encryption is told, both when it is
# saved and if one already saved is met on the message path.
No_X509_Operation_Message = 'An X.509 definition needs signing, encryption or both'

# ################################################################################################################################
# ################################################################################################################################

def keystore_from_config(config:'stranydict') -> 'Keystore':
    """ Builds a keystore out of the PEM files a WS-Security definition points to. This reads
    from disk and parses private keys, so callers on the message path use get_keystore instead.
    """

    # Our response to produce
    out = new_keystore()

    # Our own signing key with its certificate chain ..
    if signing_key_path := config.get('signing_key'):
        pem_data = Path(signing_key_path).read_bytes()
        out.signing_key = load_private_key_pem(pem_data)

    if signing_certificate_chain_path := config.get('signing_certificate_chain'):
        pem_data = Path(signing_certificate_chain_path).read_bytes()
        out.signing_certificate_chain = load_certificates_pem(pem_data)

    # .. our own decryption key - with RSA it is usually the signing key again
    # .. but a definition may keep a separate one ..
    if decryption_key_path := config.get('decryption_key'):
        pem_data = Path(decryption_key_path).read_bytes()
        out.decryption_key = load_private_key_pem(pem_data)

    # .. the other side's certificate, used both to encrypt to them and to pin their signatures ..
    if peer_certificate_path := config.get('peer_certificate'):
        pem_data = Path(peer_certificate_path).read_bytes()
        certificates = load_certificates_pem(pem_data)
        peer = certificates[0]
        out.peer_encryption_certificate = peer
        out.peer_signing_certificate = peer

    # .. and the CA certificates their signing certificates may chain up to instead of pinning.
    if trust_anchors_path := config.get('trust_anchors'):
        pem_data = Path(trust_anchors_path).read_bytes()
        out.trust_anchors = load_certificates_pem(pem_data)

    return out

# ################################################################################################################################

def _keystore_cache_key(config:'stranydict') -> 'tuple':
    """ The identity of the trust material a definition points at. The definition's id alone would
    not do, because a definition may be edited to point at different files, and the paths alone
    would not do either, because two definitions may share a file yet differ in what they do
    with it - so the key is both.
    """
    out = (
        config.get('id'),
        config.get('signing_key'),
        config.get('signing_certificate_chain'),
        config.get('decryption_key'),
        config.get('peer_certificate'),
        config.get('trust_anchors'),
    )

    return out

# ################################################################################################################################

def get_keystore(config:'stranydict') -> 'Keystore':
    """ Returns the keystore of a definition, building it on first use and reusing it afterwards.

    Building a keystore means blocking disk reads and an RSA private key parse. Doing that per
    message was the single most expensive thing on this path, and under gevent the disk read
    blocks the whole worker rather than just the greenlet doing it.
    """
    key = _keystore_cache_key(config)

    if key in _keystore_cache:
        out = _keystore_cache[key]
    else:
        out = keystore_from_config(config)
        _keystore_cache[key] = out

    return out

# ################################################################################################################################

def invalidate_keystores(definition_id:'any_') -> 'None':
    """ Forgets the keystores of one definition, which is what an edit or a delete calls for -
    the files a definition points at may have been replaced without their paths changing.
    """
    for key in list(_keystore_cache):
        if key[0] == definition_id:
            del _keystore_cache[key]

# ################################################################################################################################
# ################################################################################################################################

def _flag(config:'stranydict', name:'str', default:'bool') -> 'bool':
    """ Returns one of a definition's optional boolean flags, or what its absence means.
    """
    out = config.get(name)

    if out is None:
        out = default

    return out

# ################################################################################################################################

def _apply_username_token(envelope:'any_', config:'stranydict') -> 'None':
    """ Adds a UsernameToken with the definition's credentials, in clear text or digest form.
    """
    use_digest = _flag(config, 'use_digest', Default_Use_Digest)

    _ = add_username_token(envelope, config['username'], config['password'], use_digest)

# ################################################################################################################################

def _apply_x509(envelope:'any_', config:'stranydict') -> 'None':
    """ Signs the envelope and encrypts its body, each when the definition calls for it.
    """
    keystore = get_keystore(config)

    # Signing comes first so the signature covers the plaintext body ..
    if _flag(config, 'sign', Default_Sign):
        _ = sign(envelope, keystore)

    # .. and only then does the body turn into ciphertext.
    if _flag(config, 'encrypt', Default_Encrypt):
        encrypt_body(envelope, keystore)

# ################################################################################################################################

def _apply_saml(envelope:'any_', config:'stranydict') -> 'None':
    """ Builds a sender-vouches assertion out of the definition and places it in the security header,
    signing it first when the definition calls for it - which is what XUA-based exchanges require.
    """
    audience = config.get('audience')
    assertion = new_assertion(config['issuer'], config['subject'], audience)

    # Role and organization details travel as assertion attributes.
    if attributes := config.get('attributes'):
        for name, value in attributes.items():
            add_attribute(assertion, name, value)

    # A signed assertion is signed before it enters the header so the signature covers its final form.
    if _flag(config, 'sign', Default_Sign):
        keystore = get_keystore(config)
        _ = sign_assertion(assertion, keystore)

    add_assertion(envelope, assertion)

# ################################################################################################################################
# ################################################################################################################################

def _enforce_username_token(envelope:'any_', config:'stranydict') -> 'VerifiedSignature | None':
    """ Checks the incoming UsernameToken against the definition's credentials, in the password
    form the definition configured.
    """
    use_digest = _flag(config, 'use_digest', Default_Use_Digest)

    verify_username_token(envelope, config['username'], config['password'], use_digest)

    # A UsernameToken proves who the caller is, it does not sign anything.
    return None

# ################################################################################################################################

def _enforce_x509(envelope:'any_', config:'stranydict') -> 'VerifiedSignature | None':
    """ Decrypts the body and verifies the signature, each when the definition calls for it.
    Returns what the signature covered so the caller can bind its own reading of the message to it.
    """
    needs_decryption = _flag(config, 'encrypt', Default_Encrypt)
    needs_signature = _flag(config, 'sign', Default_Sign)

    # A signature and encryption are the two things this mode has, so a definition
    # asking for neither states no requirement for a message to meet.
    if not needs_decryption and not needs_signature:
        raise SOAPSecurityException(No_X509_Operation_Message)

    keystore = get_keystore(config)

    # Decryption comes first so the signature can be checked over the plaintext body ..
    if needs_decryption:
        decrypt_body(envelope, keystore)

    # .. and now that the body is readable, the signature over it can be verified.
    if needs_signature:
        out = verify(envelope, keystore)
    else:
        out = None

    return out

# ################################################################################################################################

def _enforce_saml(envelope:'any_', config:'stranydict') -> 'VerifiedSignature | None':
    """ Checks that the incoming message carries a valid assertion from the expected issuer.

    The issuer is a string the sender writes into the message, so on its own it proves nothing -
    the signature is what ties the assertion to an issuer that actually holds a key we trust.
    """
    assertion = get_assertion(envelope)

    # The signature is checked first. Everything below reads fields out of the assertion, and
    # until the signature verifies none of those fields mean anything - including the issuer.
    keystore = get_keystore(config)
    _ = verify_assertion(assertion, keystore)

    issuer_element = assertion.find(qname(NS.SAML2, 'Issuer'))

    if issuer_element is None:
        raise SOAPSecurityException('SAML assertion has no Issuer')

    if issuer_element.text != config['issuer']:
        raise SOAPSecurityException('SAML issuer does not match')

    validate_assertion_conditions(assertion, config.get('audience'))

    # The assertion's signature covers the assertion, not the body, so there is nothing here for
    # the caller to bind its reading of the message to.
    return None

# ################################################################################################################################
# ################################################################################################################################

# What to call to apply each mode to an outgoing envelope.
_apply_by_mode = {
    Mode.UsernameToken: _apply_username_token,
    Mode.X509:          _apply_x509,
    Mode.SAML:          _apply_saml,
}

# What to call to enforce each mode on an incoming envelope.
_enforce_by_mode = {
    Mode.UsernameToken: _enforce_username_token,
    Mode.X509:          _enforce_x509,
    Mode.SAML:          _enforce_saml,
}

# ################################################################################################################################
# ################################################################################################################################

def apply_wss(envelope:'any_', config:'stranydict') -> 'None':
    """ Applies a WS-Security definition to an outgoing envelope - whatever
    the definition's mode calls for ends up in the message's security header.
    """
    mode = config['mode']

    if apply_func := _apply_by_mode.get(mode):
        apply_func(envelope, config)

    # .. anything else is not a recognized mode.
    else:
        raise SOAPSecurityException(f'Unknown WS-Security mode `{mode}`')

# ################################################################################################################################

def enforce_wss(envelope:'any_', config:'stranydict') -> 'VerifiedSignature | None':
    """ Enforces a WS-Security definition on an incoming envelope, raising
    SOAPSecurityException when the message does not satisfy the definition's mode. Returns what
    a signature covered, when the definition's mode involves one, so the caller can check that
    the part of the message it goes on to process is the part that was verified.
    """
    mode = config['mode']

    if enforce_func := _enforce_by_mode.get(mode):
        out = enforce_func(envelope, config)

    # .. anything else is not a recognized mode.
    else:
        raise SOAPSecurityException(f'Unknown WS-Security mode `{mode}`')

    return out

# ################################################################################################################################
# ################################################################################################################################
